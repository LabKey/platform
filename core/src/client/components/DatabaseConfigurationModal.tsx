import React, { PureComponent } from 'react';
import { Button, ButtonGroup, FormControl, Modal } from 'react-bootstrap';
import { ActionURL, Ajax } from '@labkey/api';
import { DatabasePasswordRules, DatabasePasswordSettings } from "./models";

const OPTIONS_MAP = {
    Never: 'Never',
    FiveSeconds: 'Every five seconds â€” for testing',
    ThreeMonths: 'Every three months',
    SixMonths: 'Every six months',
    OneYear: 'Every twelve months',
};

interface Props {
    closeModal: Function;
    canEdit: boolean;
}

interface State {
    passwordRules: DatabasePasswordRules;
    currentSettings: DatabasePasswordSettings;
    helpLink: string;
}

export default class DatabaseConfigurationModal extends PureComponent<Props, State> {
    constructor(props) {
        super(props);
        this.state = {
            passwordRules: { Weak: '', Strong: '' },
            helpLink: null,
            currentSettings: { strength: '', expiration: '' },
        };
    }

    componentDidMount = (): void => {
        Ajax.request({
            url: ActionURL.buildURL('login', 'getDbLoginProperties'),
            method: 'GET',
            scope: this,
            failure: function(error) {
                alert('Error: ' + error);
            },
            success: function(result) {
                const response = JSON.parse(result.response);
                this.setState({ ...response });
            },
        });
    };

    handleChange = event => {
        const { name, value } = event.target;
        this.setState(prevState => ({
            ...prevState,
            currentSettings: {
                ...prevState.currentSettings,
                [name]: value,
            },
        }));
    };

    saveChanges = (): void => {
        Ajax.request({
            url: ActionURL.buildURL('login', 'SaveDbLoginProperties'),
            method: 'POST',
            jsonData: this.state.currentSettings,
            scope: this,
            failure: function(error) {
                alert('Error: ' + error);
            },
            success: function(result) {
                this.props.closeModal();
            },
        });
    };

    render() {
        const { canEdit } = this.props;
        const { currentSettings } = this.state;
        const { strength, expiration } = currentSettings;

        return (
            <Modal show={true} onHide={this.props.closeModal}>
                <Modal.Header closeButton>
                    <Modal.Title>
                        Configure Database Authentication
                    </Modal.Title>
                </Modal.Header>

                <Modal.Body>
                    <div className="database-modal__field-row">
                        <span>Password Strength:</span>

                        <span className="database-modal__field">
                            <ButtonGroup onClick={this.handleChange}>
                                <Button
                                    value="Weak"
                                    name="strength"
                                    active={strength == 'Weak'}
                                    disabled={!canEdit}>
                                    Weak
                                </Button>
                                <Button
                                    value="Strong"
                                    name="strength"
                                    active={strength == 'Strong'}
                                    disabled={!canEdit}>
                                    Strong
                                </Button>
                            </ButtonGroup>
                        </span>
                    </div>

                    {/* this.state.passwordRules.Weak is safe server-generated HTML */}
                    <div className="bold-text"> Weak </div>
                    <div>
                        <div dangerouslySetInnerHTML={{ __html: this.state.passwordRules.Weak }} />
                    </div>

                    <br/>

                    {/* this.state.passwordRules.Strong is safe server-generated HTML */}
                    <div className="bold-text"> Strong </div>
                    <div>
                        <div dangerouslySetInnerHTML={{ __html: this.state.passwordRules.Strong }} />
                    </div>

                    <br/>

                    <div className="database-modal__field-row">
                        <span>Password Expiration:</span>

                        <span className="database-modal__field">
                            {canEdit ? (
                                <FormControl
                                    componentClass="select"
                                    name="expiration"
                                    placeholder="select"
                                    onChange={this.handleChange}
                                    value={expiration}
                                >
                                    {Object.keys(OPTIONS_MAP).map((option) =>
                                        <option value={option} key={option}>
                                            {OPTIONS_MAP[option]}
                                        </option>
                                    )}
                                </FormControl>
                            ) : (
                                OPTIONS_MAP[expiration]
                            )}
                        </span>
                    </div>

                    <div className="database-modal__bottom">
                        <div className="modal__bottom-buttons">
                            <a target="_blank" href={this.state.helpLink} className="modal__help-link" rel="noopener noreferrer">
                                More about authentication
                            </a>
                            {canEdit && (
                                <Button className="labkey-button primary" onClick={this.saveChanges}>
                                    Apply
                                </Button>
                            )}
                        </div>

                        <Button className="labkey-button modal__save-button" onClick={this.props.closeModal}>
                            {canEdit ? 'Cancel' : 'Close'}
                        </Button>
                    </div>
                </Modal.Body>
            </Modal>
        );
    }
}
