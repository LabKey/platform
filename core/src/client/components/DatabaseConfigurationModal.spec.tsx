import React from 'react';

import { shallow } from 'enzyme';
import EnzymeToJson from 'enzyme-to-json';

import DatabaseConfigurationModal from './DatabaseConfigurationModal';

describe('<DatabaseConfigurationModal/>', () => {
    test('View-only', () => {
        const component = <DatabaseConfigurationModal canEdit={false} closeModal={() => {}} />;

        const wrapper = shallow(component);
        expect(EnzymeToJson(wrapper)).toMatchSnapshot();
    });

    test('Editable', () => {
        const component = <DatabaseConfigurationModal canEdit={true} closeModal={() => {}} />;

        const wrapper = shallow(component);
        expect(EnzymeToJson(wrapper)).toMatchSnapshot();
    });
});
