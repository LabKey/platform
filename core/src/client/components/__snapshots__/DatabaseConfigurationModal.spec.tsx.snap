// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DatabaseConfigurationModal/> Editable 1`] = `
<Modal
  animation={true}
  autoFocus={true}
  backdrop={true}
  bsClass="modal"
  dialogComponentClass={[Function]}
  enforceFocus={true}
  keyboard={true}
  manager={
    ModalManager {
      "add": [Function],
      "containers": Array [],
      "data": Array [],
      "handleContainerOverflow": true,
      "hideSiblingNodes": true,
      "isTopModal": [Function],
      "modals": Array [],
      "remove": [Function],
    }
  }
  onHide={[Function]}
  renderBackdrop={[Function]}
  restoreFocus={true}
  show={true}
>
  <ModalHeader
    bsClass="modal-header"
    closeButton={true}
    closeLabel="Close"
  >
    <ModalTitle
      bsClass="modal-title"
      componentClass="h4"
    >
      Configure Database Authentication
    </ModalTitle>
  </ModalHeader>
  <ModalBody
    bsClass="modal-body"
    componentClass="div"
  >
    <div
      className="database-modal__field-row"
    >
      <span>
        Password Strength:
      </span>
      <span
        className="database-modal__field"
      >
        <ButtonGroup
          block={false}
          bsClass="btn-group"
          justified={false}
          onClick={[Function]}
          vertical={false}
        >
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            disabled={false}
            name="strength"
            value="Weak"
          >
            Weak
          </Button>
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            disabled={false}
            name="strength"
            value="Strong"
          >
            Strong
          </Button>
        </ButtonGroup>
      </span>
    </div>
    <div
      className="bold-text"
    >
       Weak 
    </div>
    <div>
      <div
        dangerouslySetInnerHTML={
          Object {
            "__html": "",
          }
        }
      />
    </div>
    <br />
    <div
      className="bold-text"
    >
       Strong 
    </div>
    <div>
      <div
        dangerouslySetInnerHTML={
          Object {
            "__html": "",
          }
        }
      />
    </div>
    <br />
    <div
      className="database-modal__field-row"
    >
      <span>
        Password Expiration:
      </span>
      <span
        className="database-modal__field"
      >
        <FormControl
          bsClass="form-control"
          componentClass="select"
          name="expiration"
          onChange={[Function]}
          placeholder="select"
          value=""
        >
          <option
            key="Never"
            value="Never"
          >
            Never
          </option>
          <option
            key="FiveSeconds"
            value="FiveSeconds"
          >
            Every five seconds â€” for testing
          </option>
          <option
            key="ThreeMonths"
            value="ThreeMonths"
          >
            Every three months
          </option>
          <option
            key="SixMonths"
            value="SixMonths"
          >
            Every six months
          </option>
          <option
            key="OneYear"
            value="OneYear"
          >
            Every twelve months
          </option>
        </FormControl>
      </span>
    </div>
    <div
      className="database-modal__bottom"
    >
      <div
        className="modal__bottom-buttons"
      >
        <a
          className="modal__help-link"
          href={null}
          target="_blank"
        >
          More about authentication
        </a>
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsStyle="default"
          className="labkey-button primary"
          disabled={false}
          onClick={[Function]}
        >
          Apply
        </Button>
      </div>
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="default"
        className="labkey-button modal__save-button"
        disabled={false}
        onClick={[Function]}
      >
        Cancel
      </Button>
    </div>
  </ModalBody>
</Modal>
`;

exports[`<DatabaseConfigurationModal/> View-only 1`] = `
<Modal
  animation={true}
  autoFocus={true}
  backdrop={true}
  bsClass="modal"
  dialogComponentClass={[Function]}
  enforceFocus={true}
  keyboard={true}
  manager={
    ModalManager {
      "add": [Function],
      "containers": Array [],
      "data": Array [],
      "handleContainerOverflow": true,
      "hideSiblingNodes": true,
      "isTopModal": [Function],
      "modals": Array [],
      "remove": [Function],
    }
  }
  onHide={[Function]}
  renderBackdrop={[Function]}
  restoreFocus={true}
  show={true}
>
  <ModalHeader
    bsClass="modal-header"
    closeButton={true}
    closeLabel="Close"
  >
    <ModalTitle
      bsClass="modal-title"
      componentClass="h4"
    >
      Configure Database Authentication
    </ModalTitle>
  </ModalHeader>
  <ModalBody
    bsClass="modal-body"
    componentClass="div"
  >
    <div
      className="database-modal__field-row"
    >
      <span>
        Password Strength:
      </span>
      <span
        className="database-modal__field"
      >
        <ButtonGroup
          block={false}
          bsClass="btn-group"
          justified={false}
          onClick={[Function]}
          vertical={false}
        >
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            disabled={true}
            name="strength"
            value="Weak"
          >
            Weak
          </Button>
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            disabled={true}
            name="strength"
            value="Strong"
          >
            Strong
          </Button>
        </ButtonGroup>
      </span>
    </div>
    <div
      className="bold-text"
    >
       Weak 
    </div>
    <div>
      <div
        dangerouslySetInnerHTML={
          Object {
            "__html": "",
          }
        }
      />
    </div>
    <br />
    <div
      className="bold-text"
    >
       Strong 
    </div>
    <div>
      <div
        dangerouslySetInnerHTML={
          Object {
            "__html": "",
          }
        }
      />
    </div>
    <br />
    <div
      className="database-modal__field-row"
    >
      <span>
        Password Expiration:
      </span>
      <span
        className="database-modal__field"
      />
    </div>
    <div
      className="database-modal__bottom"
    >
      <div
        className="modal__bottom-buttons"
      >
        <a
          className="modal__help-link"
          href={null}
          target="_blank"
        >
          More about authentication
        </a>
      </div>
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="default"
        className="labkey-button modal__save-button"
        disabled={false}
        onClick={[Function]}
      >
        Close
      </Button>
    </div>
  </ModalBody>
</Modal>
`;
